util/dtgen/helper.py

```diff
@@ -4,105 +4,229 @@
 """This contains a class which is used to help generate the device tables (DT)
 files.
 """
-from typing import Dict
+from abc import ABC, abstractmethod
+from typing import Optional
 from collections import OrderedDict
 
 from topgen.lib import CEnum, CArrayMapping, Name
 from reggen.ip_block import IpBlock
 
+import logging
+
 
 def indent_text(text, indent):
     return "".join(indent + line for line in text.splitlines(True))
 
 
-class ScalarType:
-    def __init__(self, typename: object):
-        self.typename = typename
+class BaseType(ABC):
+    @abstractmethod
+    def render_type_def(self) -> str:
+        """
+        Render the type definition.
 
-    def render(self, name: Name):
-        return "{} {}".format(render_type(self.typename), name.as_snake_case())
+        Example: "struct X {int field;};"
+        """
+        pass
```

Comment:
`pass` isn't required if you have a docstring. Most lint checkers will ask you to remove it.